cubi_tk/sodar/update_samplehseet.py

```diff
@@ -0,0 +1,628 @@
+        samples = self.collect_sample_data(isa_names, sample_fields_mapping)
+
+        # add metadata values to samples
+        if self.args.metadata_all:
+            samples = samples.assign(**dict(self.args.metadata_all))
+
+        # Match sample data to ISA columns and get new study & assay dataframes
+        study_new, assay_new = self.match_sample_data_to_isa(
+            samples, isa_names, sample_fields_mapping
+        )
+        req_cols = set(REQUIRED_COLUMNS) | (
+            set(REQUIRED_IF_EXISTING_COLUMNS) & set(isa_names.keys())
+        )
+        colset = set(study.columns.tolist() + assay.columns.tolist())
+        if not req_cols.issubset(colset):
+            missing_cols = req_cols - colset
+            raise ValueError(f"Missing required columns in sample data: {', '.join(missing_cols)}")
+        if self.args.sanppy_compatible:
+            for col in study_new.columns:
+                if orig_col_name(col) in req_cols:
+                    study_new[col] = study_new[col].str.replace("-", "_")
+            for col in assay_new.columns:
+                if orig_col_name(col) in req_cols:
+                    assay_new[col] = assay_new[col].str.replace("-", "_")
+
+        # Update ISA tables with new data
+        study_final = self.update_isa_table(
+            study, study_new, self.args.overwrite, self.args.no_autofill
+        )
+        assay_final = self.update_isa_table(
+            assay, assay_new, self.args.overwrite, self.args.no_autofill
+        )
+
+        # Write new samplesheet to tsv strings, then upload via API
+        study_tsv = study_final.to_csv(
+            sep="\t", index=False, header=list(map(orig_col_name, study_final.columns))
+        )
+        assay_tsv = assay_final.to_csv(
+            sep="\t", index=False, header=list(map(orig_col_name, assay_final.columns))
+        )
+        ret = sodar_api.upload_ISA_samplesheet(
+            (isa_data["investigation"][0], investigation),
+            (isa_data["study"][0], study_tsv),
+            (isa_data["assay"][0], assay_tsv),
+        )
+        return ret
+
+    def parse_sampledata_args(self, isa_names) -> OrderedDict[str, str]:
+        """Build a dict of Sample-data filed name mappings and check consistency of provided data."""
+
+        # Some samples are defined, either through ped or sample data
+        if not self.args.ped and not self.args.sample_data:
+            raise ValueError(
+                "No sample data provided. Please provide either a ped file (`--ped`) or sample data (`-s`)."
+            )
+
+        # Base field name mapping (also used by ped)
+        sample_field_mapping = sheet_default_config[self.args.defaults]["field_column_mapping"]
+
+        # Sample data, if given, need to matche the defined fields (from default or sample_fields)
+        if self.args.sample_data:
+            if self.args.sample_fields:
+                sample_field_mapping.update(
+                    OrderedDict(
```

Comment:
Which minium python version does cubi-tk require? Since python 3.7, dicts are guaranteed to be ordered by insertion order.
In that case, can probably be simplified to something along the lines of
`{k: (v or k) for (k, _, v) in map(lambda s: s.partition("="), self.args.sample_fields)}` (but not sure if that is more readable)

Reply:
ah I didn't know that about dictionaries. I think cubi-tk should have at least 3.7, so a few things here can probably be simplified